bionic/dagviz.py

```diff
@@ -4,16 +4,53 @@
 is also required.
 """
 
+import warnings
+from pathlib import Path
 from collections import defaultdict
-from io import BytesIO
+from io import BytesIO, IOBase
 
-from .optdep import import_optional_dependency
+from .optdep import import_optional_dependency, oneline
 module_purpose = 'rendering the flow DAG'
 hsluv = import_optional_dependency('hsluv', purpose=module_purpose)
 pydot = import_optional_dependency('pydot', purpose=module_purpose)
 Image = import_optional_dependency('PIL.Image', purpose=module_purpose)
 
 
+class FlowImage:
+    def __init__(self, pydot_graph):
+        '''
+        Given a pydot graph object, create Pillow Image or SVG represented as XML text.
+        '''
+        self.pil_image = Image.open(BytesIO(pydot_graph.create_png()))
+        self.xml_text = pydot_graph.create_svg()
+
+    def save(self, fp, format=None, **params):
+        '''Save flow visualization under the given filename. For SVG, must pass a filepath and does not
```

Comment:
The `format` parameter is specifically for formats supported by PIL. (for example: `quality`, which doesn't have a clear analog for the built in python save). I changed `format` >`pil_format` to clarify this.

Regarding saving SVG to a file object, we could probably add a 2nd format parameter like `file_format` in addition to `pil_format` to be able to specify this? I understand trying to make the API as simple as possible, but it's not clear to me how many teams are saving these visualizations to file objects? Feel free to push back though

Reply:
Sorry, I'm not sure I'm following what you mean about the `format` parameter. It just tells PIL which of the supported formats you want, right? `quality` would be something you pass to `params`, which is indeed format-specific.

Given that we're using PIL as a starting point for our API (and given that we want to be backwards-compatible) it seems like the simplest API we can present is "just like PIL, but we support one extra format called `'svg'`".

In particular, our current API (which is just PIL) already supports the following:
* `image.save('myfile.png')`
* `image.save('myfile.whatever', format='png')`
* `image.save(some_file_object, format='png')`

If we're going to continue supporting all of those, then why not support all the equivalents for SVG?
* `image.save('myfile.svg')`
* `image.save('myfile.whatever', format='svg')`
* `image.save(some_file_object, format='svg')`

If we do something different, then the user has to learn and remember that there's a different set of rules for SVG vs all other formats, and what they are. In this case, that seems to me to outweigh the extra lines code needed to support PIL's conventions. (In fact, I suspect it would take even more lines of extra documentation to explain our more complex API.)